Source/Scene/ModelExperimental/InstancingPipelineStage.js

```diff
@@ -291,114 +549,295 @@ function getInstanceTransformsTypedArray(instances, count, renderResources) {
+
+  return transforms;
+}
+
+function getInstanceTranslationsAsCartesian3s(translationAttribute, count) {
+  const translations = new Array(count);
+  const translationTypedArray = translationAttribute.packedTypedArray;
 
-    renderResources.instancingTranslationMax = instancingTranslationMax;
-    renderResources.instancingTranslationMin = instancingTranslationMin;
+  for (let i = 0; i < count; i++) {
+    translations[i] = new Cartesian3(
+      translationTypedArray[i * 3],
+      translationTypedArray[i * 3 + 1],
+      translationTypedArray[i * 3 + 2]
+    );
   }
 
-  return transformsTypedArray;
+  return translations;
 }
 
-function processFeatureIdAttributes(
+function createVertexBuffer(typedArray, frameState) {
+  const buffer = Buffer.createVertexBuffer({
+    context: frameState.context,
+    typedArray: typedArray,
+    usage: BufferUsage.STATIC_DRAW,
+  });
+
+  // Destruction of resources allocated by the ModelExperimental
+  // is handled by ModelExperimental.destroy().
+  buffer.vertexArrayDestroyable = false;
+
+  return buffer;
+}
+
+function processTransformAttributes(
   renderResources,
   frameState,
   instances,
-  instancingVertexAttributes
+  instancingVertexAttributes,
+  use2D
 ) {
-  const attributes = instances.attributes;
-  const model = renderResources.model;
+  const translationAttribute = ModelExperimentalUtility.getAttributeBySemantic(
+    instances,
+    InstanceAttributeSemantic.TRANSLATION
+  );
+
+  let translationMax;
+  let translationMin;
+  if (defined(translationAttribute)) {
+    translationMax = translationAttribute.max;
+    translationMin = translationAttribute.min;
+  }
+
+  const rotationAttribute = ModelExperimentalUtility.getAttributeBySemantic(
+    instances,
+    InstanceAttributeSemantic.ROTATION
+  );
+
   const shaderBuilder = renderResources.shaderBuilder;
+  const count = instances.attributes[0].count;
+  const useMatrices =
+    defined(rotationAttribute) ||
+    !defined(translationMax) ||
+    !defined(translationMin);
 
-  // Load Feature ID vertex attributes. These are loaded as typed arrays in GltfLoader
-  // because we want to expose the instance feature ID when picking.
-  for (let i = 0; i < attributes.length; i++) {
-    const attribute = attributes[i];
-    if (attribute.semantic !== InstanceAttributeSemantic.FEATURE_ID) {
-      continue;
+  const statistics = renderResources.model.statistics;
+
+  // Packed typed arrays are omitted from statistics because they don't
+  // necessarily correspond to the size of the GPU buffer containing
+  // their data. It's also difficult to track which typed arrays have
+  // already been counted.
+  const hasCpuCopy = false;
+
+  let transforms;
+  if (useMatrices) {
+    shaderBuilder.addDefine("HAS_INSTANCE_MATRICES");
+    const attributeString = "Transform";
+
+    transforms = getInstanceTransformsAsMatrices(
+      instances,
+      count,
+      renderResources,
+      use2D
```

Comment:
Without any context here: This `use2D` does not seem to be used (i.e. the function signature does not contain this argument)

Reply:
Thanks for the catch @gpu , I'll fix that shortly